crypto/rsa/rsa_sp800_56b_gen.c

```diff
@@ -0,0 +1,327 @@
+                       RSA_R_INVALID_MODULUS);
+                return 0;
+            }
+        } else if (strength == 128) {
+            if (nbits != 3072) {
+                RSAerr(RSA_F_RSA_SP800_56B_VALIDATE_STRENGTH,
+                       RSA_R_INVALID_MODULUS);
+                return 0;
+            }
+        } else {
+            RSAerr(RSA_F_RSA_SP800_56B_VALIDATE_STRENGTH,
+                   RSA_R_INVALID_STRENGTH);
+            return 0;
+        }
+    }
+    return 1;
+}
+
+/*
+ * SP800-56Br1
+ * 6.3.1.1 rsakpg1 - basic (Steps 3-4)
+ * 6.3.1.3 rsakpg1 - crt   (Step 5)
+ *
+ * Using p & q, calculate other required parameters such as n, d.
+ * as well as the CRT parameters dP, dQ, qInv.
+ *
+ * Note:
+ *   There is a small chance that the generated d will be too small.
+ * Returns -1 = error,
+ *          0 = d is too small,
+ *          1 = success.
+ */
+int rsa_sp800_56b_derive_params_from_pq(RSA *rsa, int nbits,
+                                        const BIGNUM *e, BN_CTX *ctx)
+{
+    int ret = -1;
+    BN_CTX *ctx2 = NULL;
+    BIGNUM *p1, *q1, *lcm, *p1q1, *gcd;
+
+    BN_CTX_start(ctx);
+    p1 = BN_CTX_get(ctx);
+    q1 = BN_CTX_get(ctx);
+    lcm = BN_CTX_get(ctx);
+    p1q1 = BN_CTX_get(ctx);
+    gcd = BN_CTX_get(ctx);
+    if (gcd == NULL)
+        goto err;
+
+    /* LCM((p-1, q-1)) */
+    if (rsa_get_lcm(ctx, rsa->p, rsa->q, lcm, gcd, p1, q1, p1q1) != 1)
+        goto err;
+
+    ctx2 = BN_CTX_new();
+    if (ctx2 == NULL)
+        goto err;
+
+    /* copy e */
+    BN_free(rsa->e);
+    rsa->e = BN_dup(e);
+    if (rsa->e == NULL)
+        goto err;
+
+    BN_clear_free(rsa->d);
+    /* (Step 3) d = (e^-1) mod (LCM(p-1, q-1)) */
+    rsa->d = BN_secure_new();
+    if (rsa->d == NULL || BN_mod_inverse(rsa->d, e, lcm, ctx2) == NULL)
+        goto err;
+
+    /* (Step 3) return an error if d is too small */
+    if (BN_num_bits(rsa->d) <= (nbits >> 1)) {
+        ret = 0;
+        goto err;
+    }
+
+    /* (Step 4) n = pq */
+    if (rsa->n == NULL)
+        rsa->n = BN_new();
+    if (rsa->n == NULL || !BN_mul(rsa->n, rsa->p, rsa->q, ctx))
+        goto err;
+
+    /* (Step 5a) dP = d mod (p-1) */
+    if (rsa->dmp1 == NULL)
+        rsa->dmp1 = BN_new();
+    if (rsa->dmp1 == NULL || !BN_mod(rsa->dmp1, rsa->d, p1, ctx))
+        goto err;
+
+    /* (Step 5b) dQ = d mod (q-1) */
+    if (rsa->dmq1 == NULL)
+        rsa->dmq1 = BN_secure_new();
+    if (rsa->dmq1 == NULL || !BN_mod(rsa->dmq1, rsa->d, q1, ctx))
+        goto err;
+
+    /* (Step 5c) qInv = (inverse of q) mod p */
+    BN_free(rsa->iqmp);
+    rsa->iqmp = BN_secure_new();
+    if (rsa->iqmp == NULL
+            || BN_mod_inverse(rsa->iqmp, rsa->q, rsa->p, ctx2) == NULL)
+        goto err;
+
+    ret = 1;
+err:
+    if (ret != 1) {
+        BN_free(rsa->e);
+        BN_free(rsa->d);
+        BN_free(rsa->n);
+        BN_free(rsa->iqmp);
+        BN_free(rsa->dmq1);
+        BN_free(rsa->dmp1);
```

Comment:
Doesn't that leave hanging references in the RSA object?

Reply:
Good point. Will set all to NULL.